### ISSUE TITLE
 Quest v0.2 / G6PD: Extraction of FHIR resources & Generate Observation Resources

### ISSUE THREAD

**Issue by mwkhan1983** [opensrp/fhircore#599]
We need to decide on the following approaches based on the current requirements and for RIM to work 

**Step 1** 
- [x] Created and save a [QuestionnaireResponse](https://www.hl7.org/fhir/questionnaireresponse.html) 
- [x] Extract resources for RIM computation  - extraction of Patient , Observation and Pregnancy condition (If selected)

Sample resources to be extracted and generated 

1. Pregnancy Condition (_NB. Extract if, pregnancy was checked during registration_)

```
{
    "resourceType": "Condition",
    "clinicalStatus": {
        "coding": [
            {
                "system": "http://terminology.hl7.org/CodeSystem/condition-clinical",
                "code": "active"
            }
        ]
    },
    "verificationStatus": {
        "coding": [
            {
                "system": "http://terminology.hl7.org/CodeSystem/condition-ver-status",
                "code": "confirmed"
            }
        ]
    },
    "code": {
        "coding": [
            {
                "system": "http://snomed.info/sct",
                "code": "77386006",
                "display": "Pregnancy"
            }
        ],
        "text": "Pregnancy"
    },
    "subject": {
        "reference": "Patient/XXX"
    }
}
```
2. Haem Observation
```
{
  "resourceType": "Observation",
  "status": "registered",
  "code": {
      "coding": [
          {
              "system": "http://snomed.info/sct",
              "code": "259695003",
              "display": "Hemoglobin [Mass/volume] in Blood"
          }
      ],
      "text": "Haemoglobin"
  },
  "subject": {
      "reference": "Patient/XXX"
  },
  "effectivePeriod": {
      "start": "2013-04-05T10:30:10+01:00",
      "end": "2013-04-05T10:30:10+01:00"
  },
  "issued": "2013-04-05T15:30:10+01:00",
  "valueQuantity": {
      "value": 8.5,
      "unit": "g/dL",
      "system": "http://snomed.info/sct",
      "code": "259695003"
  }
}
```

3. G6PD Observation 
```
{
  "resourceType": "Observation",
  "status": "registered",
  "code": {
      "coding": [
          {
              "system": "http://snomed.info/sct",
              "code": "86859003",
              "display": "G6PD in Blood"
          }
      ],
      "text": "G6PD"
  },
  "subject": {
      "reference": "Patient/XXX"
  },
  "effectivePeriod": {
      "start": "2013-04-05T10:30:10+01:00",
      "end": "2013-04-05T10:30:10+01:00"
  },
  "issued": "2013-04-05T15:30:10+01:00",
  "valueQuantity": {
      "value": 15,
      "unit": "μg Hb/",
      "system": "http://unitsofmeasure.org",
      "code": "μg Hb/"
  }
}
```

**Step  2** To be done on a separate ticket (https://github.com/opensrp/fhircore/issues/906)
- [x] Create a [ServiceRequest](https://www.hl7.org/fhir/servicerequest.html) and a [DiagnosticReport](https://www.hl7.org/fhir/diagnosticreport.html) 
- [ ] Capture Device Serial number and save to the [Device](https://www.hl7.org/fhir/device.html) resource 
- [ ] Link [DeviceDefinition](http://www.hl7.org/fhir/devicedefinition.html) while saving the [Device](http://hl7.org/fhir/device.html) details 
- [ ] Determine how to store data post RIM computation is done, will need to at least have the following resources, **Patient**, **Practitioner**, **Device**, **Observation** and **DiagnosticReport**  created to query a patient record  via https://turn-fhir.smartregister.org/fhir/Patient/XXX/$everything where XXX is PatientID  
- [x] Push data to RIM via [Observation](https://www.hl7.org/fhir/observation.html) resource 

Sample resources to be extracted and generated 

1. Service Request 

```
{
  "resourceType": "ServiceRequest",
  "status": "active",
  "intent": "original-order",
  "subject": {
    "reference": "Patient/3536"
  },
  "basedOn": [
    {
      "reference": "QuestionnaireResponse/XXX"
    }
  ],
  "occurrenceDateTime": "2021-05-12T16:16:00-07:00",
  "reasonCode": [
    {
      "coding": [
        {
          "system": "http://snomed.info/sct",
          "code": "86859003",
          "display": "Glucose-6-phosphate dehydrogenase deficiency anaemia analysis"
        }
      ],
      "text": "G6PD Test"
    }
  ]
}
```

**Comment by mwkhan1983** [comment]
Peter: is the Observation resource (below) the only thing being extracted?

**Comment by mwkhan1983** [comment]
We will be generating observation request with reference to patient and practitioner resources.

**Comment by mwkhan1983** [comment]
Peter: Is anything else being generated from the G6PD test results questionnaire response?

**Comment by pld** [comment]
Note we need to do this with no custom coding, a StructureMap linked to the Questionnaire seems like an approach that will get us there

**Comment by pld** [comment]
How we do the parts around DeviceDefinition and Device are still unclear because we need to discuss the use case requirements here and liase with the design team, I suggest we move that to a new issue to unblock this one.

I see overlap between the two options, and I think regardless to close and unblock this issue we can scope it as:

1. write a StructureMap to apply after form submission to go from QuestionnaireResponse to
    - Observation
    - DiagnosticReport
    - ServiceRequest
    -  (also continue to write the the raw QuestionnaireResponse to the DB for sync as well)

How about that?

**Comment by f-odhiambo** [comment]
> How we do the parts around DeviceDefinition and Device are still unclear because we need to discuss the use case requirements here and liase with the design team, I suggest we move that to a new issue to unblock this one.
> 
> I see overlap between the two options, and I think regardless to close and unblock this issue we can scope it as:
> 
> 1. write a StructureMap to apply after form submission to go from QuestionnaireResponse to
>    
>    * Observation
>    * DiagnosticReport
>    * ServiceRequest
>    * (also continue to write the the raw QuestionnaireResponse to the DB for sync as well)
> 
> How about that?

This sounds good based on the updated flow [here](https://docs.google.com/drawings/d/1R5iG63u5z2K9omPSFSwoyc5FDMtK_ws51BhcC4DqYRk/edit)

**Comment by f-odhiambo** [comment]
Sample resources
1. Service Request 

```
{
  "resourceType": "ServiceRequest",
  "status": "active",
  "intent": "original-order",
  "subject": {
    "reference": "Patient/3536"
  },
  "basedOn": [
    {
      "reference": "QuestionnaireResponse/XXX"
    }
  ],
  "occurrenceDateTime": "2021-05-12T16:16:00-07:00",
  "reasonCode": [
    {
      "coding": [
        {
          "system": "http://snomed.info/sct",
          "code": "86859003",
          "display": "Glucose-6-phosphate dehydrogenase deficiency anaemia analysis"
        }
      ],
      "text": "G6PD Test"
    }
  ]
}
```
2. Pregnancy Condition 

```
{
    "resourceType": "Condition",
    "clinicalStatus": {
        "coding": [
            {
                "system": "http://terminology.hl7.org/CodeSystem/condition-clinical",
                "code": "active"
            }
        ]
    },
    "verificationStatus": {
        "coding": [
            {
                "system": "http://terminology.hl7.org/CodeSystem/condition-ver-status",
                "code": "confirmed"
            }
        ]
    },
    "code": {
        "coding": [
            {
                "system": "http://snomed.info/sct",
                "code": "77386006",
                "display": "Pregnancy"
            }
        ],
        "text": "Pregnancy"
    },
    "subject": {
        "reference": "Patient/XXX"
    }
}
```
3. Haem Observation
```
{
  "resourceType": "Observation",
  "id": "115",
  "meta": {
      "versionId": "46",
      "lastUpdated": "2021-11-14T19:08:46.881+00:00",
      "source": "#yZn2obJz7bjcOEFn"
  },
  "status": "registered",
  "code": {
      "coding": [
          {
              "system": "http://snomed.info/sct",
              "code": "259695003",
              "display": "Hemoglobin [Mass/volume] in Blood"
          }
      ],
      "text": "Haemoglobin"
  },
  "subject": {
      "reference": "Patient/XXX"
  },
  "effectivePeriod": {
      "start": "2013-04-05T10:30:10+01:00",
      "end": "2013-04-05T10:30:10+01:00"
  },
  "issued": "2013-04-05T15:30:10+01:00",
  "valueQuantity": {
      "value": 8.5,
      "unit": "g/dL",
      "system": "http://snomed.info/sct",
      "code": "259695003"
  }
}
```

4. G6PD Observation 
```
{
  "resourceType": "Observation",
  "id": "114",
  "meta": {
      "versionId": "15",
      "lastUpdated": "2021-11-14T19:08:48.374+00:00",
      "source": "#H8jhdzsh5onDYmu8"
  },
  "status": "registered",
  "code": {
      "coding": [
          {
              "system": "http://snomed.info/sct",
              "code": "86859003",
              "display": "G6PD in Blood"
          }
      ],
      "text": "G6PD"
  },
  "subject": {
      "reference": "Patient/XXX"
  },
  "effectivePeriod": {
      "start": "2013-04-05T10:30:10+01:00",
      "end": "2013-04-05T10:30:10+01:00"
  },
  "issued": "2013-04-05T15:30:10+01:00",
  "valueQuantity": {
      "value": 15,
      "unit": "μg Hb/",
      "system": "http://unitsofmeasure.org",
      "code": "μg Hb/"
  }
}
```

**Comment by f-odhiambo** [comment]
@maimoonak  QA on this could also include checking the required input fields referenced in ticket #806

**Comment by maimoonak** [comment]
- Additional encounter created so that it can combine multiple results under one enc
- quantity unit/code in above Obs Haem is not correct. Changed to g/dL
- Issued not created into Obs as 'Instant' can not be created. (effective created)
- Codes copied from test results instead of using above otherwise CQL would not run
- Three obs of questionnaire mapped 2 above mentioned + 1 result type
- Different mapping for G6PD which exists in questionnaire

**Comment by maimoonak** [comment]
**Patient**
`
{"resourceType":"Patient","id":"9d55a44d-6a95-4213-ab71-8a5ecb835e2a","meta":{"tag":[{"system":"http://fhir.ona.io","code":"000003","display":"G6PD Patient"}]},"extension":[{"url":"http://hl7.org/fhir/us/core/StructureDefinition/us-core-ethnicity","valueCoding":{"code":"black","display":"Black"}},{"url":"http://hl7.org.nz/fhir/StructureDefinition/current-occupation","valueCoding":{"code":"employed","display":"Employed"}}],"identifier":[{"value":"3456"}],"active":true,"name":[{"family":"Hadi","given":["Bareera"]}],"telecom":[{"system":"phone","value":"4688755"}],"gender":"female","birthDate":"1998-12-14","address":[{"use":"home","type":"physical","city":"NAIROBI","country":"Kenya"}],"maritalStatus":{"coding":[{"code":"married","display":"Married"}]},"managingOrganization":{"reference":"Organization/105"}}
`

**Condition**
`
{"resourceType":"Condition","id":"3ed87055-e972-482b-af16-9ca7bd0c5690","clinicalStatus":{"coding":[{"system":"http://terminology.hl7.org/CodeSystem/condition-clinical","code":"active"}]},"verificationStatus":{"coding":[{"system":"http://terminology.hl7.org/CodeSystem/condition-ver-status","code":"confirmed"}]},"code":{"coding":[{"system":"http://snomed.info/sct","code":"77386006"}],"text":"Pregnant"},"subject":{"reference":"Patient/9d55a44d-6a95-4213-ab71-8a5ecb835e2a"},"recordedDate":"2021-12-14T11:45:45+05:00"}
`

**Encounter**
`
{"resourceType":"Encounter","id":"2bcb7ad7-4b14-48cf-bfb4-e858b3596e5a","status":"finished","class":{"system":"http://terminology.hl7.org/CodeSystem/v3-ActCode","code":"AMB"},"serviceType":{"coding":[{"system":"http://terminology.hl7.org/CodeSystem/service-type","code":"581"}]},"subject":{"reference":"Patient/9d55a44d-6a95-4213-ab71-8a5ecb835e2a"}}
`

**Observation Result Type**
`
{"resourceType":"Observation","id":"#a50c9a7b-9220-4c13-9859-9ba83e539436","status":"registered","category":[{"coding":[{"system":"http://terminology.hl7.org/CodeSystem/observation-category","code":"laboratory"}]}],"code":{"coding":[{"system":"http://fhir.ona.io","code":"000001"}],"text":"G6PD Result Type"},"subject":{"reference":"Patient/9d55a44d-6a95-4213-ab71-8a5ecb835e2a"},"encounter":{"reference":"Encounter/2bcb7ad7-4b14-48cf-bfb4-e858b3596e5a"},"effectivePeriod":{"start":"2021-12-14T11:47:33+05:00","end":"2021-12-14T11:47:33+05:00"},"valueCodeableConcept":{"coding":[{"system":"https://www.snomed.org","code":"410680006","display":"Number"}]}}
`

**Observation G6PD**
`
{"resourceType":"Observation","id":"#1befd1ce-6d8b-4611-a171-f3fe503e369b","status":"registered","category":[{"coding":[{"system":"http://terminology.hl7.org/CodeSystem/observation-category","code":"laboratory"}]}],"code":{"coding":[{"system":"http://snomed.info/sct","code":"86859003"}],"text":"G6PD"},"subject":{"reference":"Patient/9d55a44d-6a95-4213-ab71-8a5ecb835e2a"},"encounter":{"reference":"Encounter/2bcb7ad7-4b14-48cf-bfb4-e858b3596e5a"},"effectivePeriod":{"start":"2021-12-14T11:47:33+05:00","end":"2021-12-14T11:47:33+05:00"},"valueQuantity":{"value":4.0,"unit":"U/g Hb","system":"http://unitsofmeasure.org","code":"U/g Hb"}}
`

**Observation Haemoglobin**
`
{"resourceType":"Observation","id":"#35517341-a08c-4021-9c1f-d4daa827641e","status":"registered","category":[{"coding":[{"system":"http://terminology.hl7.org/CodeSystem/observation-category","code":"laboratory"}]}],"code":{"coding":[{"system":"http://snomed.info/sct","code":"259695003"}],"text":"G6PD"},"subject":{"reference":"Patient/9d55a44d-6a95-4213-ab71-8a5ecb835e2a"},"encounter":{"reference":"Encounter/2bcb7ad7-4b14-48cf-bfb4-e858b3596e5a"},"effectivePeriod":{"start":"2021-12-14T11:47:33+05:00","end":"2021-12-14T11:47:33+05:00"},"valueQuantity":{"value":13.0,"unit":"g/dL","system":"http://unitsofmeasure.org","code":"g/dL"}}
`

**Comment by maimoonak** [comment]
**Step 2 To be done on a separate ticket** Kindly link the ticket for the rest of the work. The CQL Condition extraction is under progress with #779

**Comment by mwkhan1983** [comment]
@maimoonak created new ticket for step 2 implementation #906

**Comment by f-odhiambo** [comment]
Marking this as DONE @maimoonak
